main/src/main/scala/sbt/internal/server/NetworkChannel.scala

```diff
@@ -71,15 +71,24 @@ final class NetworkChannel(val name: String, connection: Socket, state: State) e
     }
 
   def onCommand(command: CommandMessage): Unit = command match {
-    case x: ExecCommand => onExecCommand(x)
+    case x: ExecCommand      => onExecCommand(x)
+    case x: GetSettingReq[_] => onGetSettingReq(x)
   }
 
   private def onExecCommand(cmd: ExecCommand) =
     append(Exec(cmd.commandLine, cmd.execId orElse Some(Exec.newExecId), Some(CommandSource(name))))
 
+  private def onGetSettingReq[A](req: GetSettingReq[A]) = {
+    val rsp = GetSettingRsp((Project structure state).data.get(req.settingKey.scope, req.settingKey.key))
+    StandardMain.exchange publishEventMessage rsp
+  }
+
   def shutdown(): Unit = {
     println("Shutting down client connection")
     running.set(false)
     out.close()
   }
 }
+
+final case class GetSettingReq[A](settingKey: SettingKey[A]) extends CommandMessage
```

Comment:
1. Probably move to protocol subproject. 
2. Rename to `SettingQuery`?
3. Avoid case class.